src/Init/Lean/Elab/Quotation.lean

```diff
@@ -6,67 +6,85 @@ Authors: Sebastian Ullrich
+
+instance Name.HasQuote : HasQuote Name := ⟨quoteName⟩
+
+private def appN (fn : Syntax) (args : Array Syntax) : Syntax :=
+args.foldl (fun fn arg => Unhygienic.run `(%%fn %%arg)) fn
+
+instance Prod.HasQuote {α β : Type} [HasQuote α] [HasQuote β] : HasQuote (α × β) :=
+⟨fun ⟨a, b⟩ => Unhygienic.run `(_root_.Prod.mk %%(Lean.HasQuote.quote a) %%(Lean.HasQuote.quote b))⟩
+
+private def quoteList {α : Type} [HasQuote α] : List α → Syntax
+| [] =>      Unhygienic.run `(_root_.List.nil)
+| (x::xs) => Unhygienic.run `(_root_.List.cons %%(Lean.HasQuote.quote x) %%(quoteList xs))
 
-private def quoteList : List Syntax → Syntax
-| [] => const `List.nil
-| (x::xs) => appN (const `List.cons) #[x, quoteList xs]
+instance List.HasQuote {α : Type} [HasQuote α] : HasQuote (List α) := ⟨quoteList⟩
 
-private def quoteArray : Array Syntax → Syntax
-| xs => app (const `List.toArray) $ quoteList xs.toList
+instance Array.HasQuote {α : Type} [HasQuote α] : HasQuote (Array α) :=
+⟨fun xs => let stx := quote xs.toList; Unhygienic.run `(_root_.List.toArray %%stx)⟩
 
-private partial def quote (env : Environment) (msc : Syntax) : Syntax → Syntax
-| Syntax.ident info rawVal val preresolved =>
+namespace Elab
+namespace Term
+
+private partial def quoteSyntax {m : Type → Type} [Monad m] [MonadQuotation m] (env : Environment) : Nat → Syntax → m Syntax
+| _, Syntax.ident info rawVal val preresolved => do
   -- TODO: pass scope information
   let ns := Name.anonymous;
   let openDecls := [];
   let preresolved := resolveGlobalName env ns openDecls val <|> preresolved;
-  --`(Syntax.ident none (String.toSubstring %(rawVal.toString)) (Name.mkNumeral %val %msc))
-  appN (const `Lean.Syntax.ident) #[
-    const `Option.none,
-    app (const `String.toSubstring) (mkStxStrLit rawVal.toString),
-    --appN (const `Lean.Name.mkNumeral) #[quoteName val, msc]
-    -- TODO: hygiene
-    quoteName val,
-    quoteList $ preresolved.map $ fun ⟨n, ss⟩ => appN (const `Prod.mk) #[quoteName n, quoteList $ ss.map mkStxStrLit]
-  ]
-| Syntax.node `Lean.Parser.Term.antiquot args => args.get! 1
-| Syntax.node k args =>
-  --`(Syntax.node %k %args...)
-  let args := quoteArray $ args.map quote;
-  appN (const `Lean.Syntax.node) $ #[quoteName k, args]
-| Syntax.atom info val => --`(Syntax.atom none %val)
-  appN (const `Lean.Syntax.atom) #[
-    const `Option.none,
-    mkStxStrLit val
-  ]
-| Syntax.missing => unreachable!
-
--- TODO: hygiene
-def stxQuot.expand (env : Environment) (stx : Syntax) : Syntax :=
--- `(do msc ← getCurMacroScope; pure %(quote `(msc) quoted))
-app (const `HasPure.pure) $ quote env Syntax.missing $ stx.getArg 1
+  let val := quote val;
+  -- `scp` is bound in stxQuot.expand
+  val ← `(Lean.addMacroScope %%val scp);
+  let args := quote preresolved;
+  `(Lean.Syntax.ident Option.none (String.toSubstring %%(Lean.mkStxStrLit (HasToString.toString rawVal))) %%val %%args)
+| 0, Syntax.node `Lean.Parser.Term.antiquot args => pure $ args.get! 1
```

Comment:
Please add some comments here. I only managed to understand this code after I read the example `stxQuot.lean`.
Suggestion:
- Add examples as a docstring for this method
- Explain how `lvl` is used. The examples are super useful for making this part clear.

Reply:
I added comments about `lvl` in general. I will add examples as soon as I've figured out what should actually happen inside nested quotations, because I don't think the current behavior makes sense. This is related to the "brain teaser" I left on Zulip. I can completely remove `lvl` for now if you want to until we've figured it out.